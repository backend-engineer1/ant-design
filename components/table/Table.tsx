import * as React from 'react';
import { Summary, EXPAND_COLUMN } from 'rc-table';
import Column from './Column';
import ColumnGroup from './ColumnGroup';
import InternalTable from './InternalTable';
import type { TableProps } from './InternalTable';
import type { RefTable } from './interface';
import {
  SELECTION_ALL,
  SELECTION_COLUMN,
  SELECTION_INVERT,
  SELECTION_NONE,
} from './hooks/useSelection';

function Table<RecordType extends object = any>(
  props: TableProps<RecordType>,
  ref: React.Ref<HTMLDivElement>,
) {
  const renderTimesRef = React.useRef(0);
  renderTimesRef.current += 1;

  return <InternalTable<RecordType> {...props} ref={ref} _renderTimes={renderTimesRef.current} />;
}

const ForwardTable = React.forwardRef(Table) as any as RefTable & {
  SELECTION_COLUMN: typeof SELECTION_COLUMN;
  EXPAND_COLUMN: typeof EXPAND_COLUMN;
  SELECTION_ALL: 'SELECT_ALL';
  SELECTION_INVERT: 'SELECT_INVERT';
  SELECTION_NONE: 'SELECT_NONE';
  Column: typeof Column;
  ColumnGroup: typeof ColumnGroup;
  Summary: typeof Summary;
};

ForwardTable.SELECTION_COLUMN = SELECTION_COLUMN;
ForwardTable.EXPAND_COLUMN = EXPAND_COLUMN;
ForwardTable.SELECTION_ALL = SELECTION_ALL;
ForwardTable.SELECTION_INVERT = SELECTION_INVERT;
ForwardTable.SELECTION_NONE = SELECTION_NONE;
ForwardTable.Column = Column;
ForwardTable.ColumnGroup = ColumnGroup;
ForwardTable.Summary = Summary;

export default ForwardTable;
