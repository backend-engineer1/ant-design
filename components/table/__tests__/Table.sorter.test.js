import React from 'react';
import { render, mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import Table from '..';

describe('Table.sorter', () => {
  const sorterFn = (a, b) => a.name[0].charCodeAt() - b.name[0].charCodeAt();

  const column = {
    title: 'Name',
    dataIndex: 'name',
    sorter: sorterFn,
  };

  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  function createTable(props) {
    return (
      <Table
        columns={[column]}
        dataSource={data}
        pagination={false}
        {...props}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('TableRow').map(row => row.props().record.name);
  }

  it('renders sorter icon correctly', () => {
    const wrapper = render(createTable());
    expect(renderToJson(wrapper.find('thead'))).toMatchSnapshot();
  });

  it('sort records', () => {
    const wrapper = mount(createTable());

    wrapper.find('.ant-table-column-sorter-up').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);

    wrapper.find('.ant-table-column-sorter-down').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  it('can be controlled by sortOrder', () => {
    const wrapper = mount(createTable({
      columns: [{ ...column, sortOrder: 'ascend' }],
    }));
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
  });

  it('fires change event', () => {
    const handleChange = jest.fn();
    const wrapper = mount(createTable({ onChange: handleChange }));

    wrapper.find('.ant-table-column-sorter-up').simulate('click');

    const sorter = handleChange.mock.calls[0][2];
    expect(sorter.column.dataIndex).toBe('name');
    expect(sorter.order).toBe('ascend');
    expect(sorter.field).toBe('name');
    expect(sorter.columnKey).toBe('name');
  });
});
