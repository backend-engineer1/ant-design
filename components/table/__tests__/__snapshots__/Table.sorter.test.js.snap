// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.sorter renders sorter icon correctly 1`] = `
<thead
  class="ant-table-thead"
>
  <tr>
    <th
      class="ant-table-column-has-actions ant-table-column-has-sorters"
    >
      <div
        class="ant-table-column-sorters"
        title="Sort"
      >
        Name
        <div
          class="ant-table-column-sorter"
        >
          <i
            class="anticon anticon-caret-up ant-table-column-sorter-up off"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="caret-up"
              fill="currentColor"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M76.3 632.8l405.1-489.1a42.53 42.53 0 0 1 63.4 0l403.1 489.1c24.4 27.4 5 70.8-31.7 70.8H107.9c-36.6 0-56.1-43.4-31.6-70.8z"
              />
            </svg>
          </i>
          <i
            class="anticon anticon-caret-down ant-table-column-sorter-down off"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="caret-down"
              fill="currentColor"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M76.3 392.3l403.1 489.1a42.53 42.53 0 0 0 63.4 0l405.1-489.1c24.4-27.4 5-70.8-31.7-70.8H108.1c-36.8 0-56.3 43.4-31.8 70.8z"
              />
            </svg>
          </i>
        </div>
      </div>
    </th>
  </tr>
</thead>
`;
